linkerd/meshtls/boring/src/creds.rs

```diff
@@ -0,0 +1,210 @@
+    base: Arc<BaseCreds>,
+    certs: Option<Certs>,
+}
+
+struct BaseCreds {
+    roots: Vec<X509>,
+    key: PKey<Private>,
+}
+
+struct Certs {
+    leaf: X509,
+    intermediates: Vec<X509>,
+}
+
+pub(crate) type CredsRx = watch::Receiver<Creds>;
+
+type CredsTx = watch::Sender<Creds>;
+
+// === impl Creds ===
+
+impl From<Arc<BaseCreds>> for Creds {
+    fn from(base: Arc<BaseCreds>) -> Self {
+        Self { base, certs: None }
+    }
+}
+
+impl Creds {
+    // TODO(ver) Specify certificate types, signing algorithms, cipher suites..
+    pub(crate) fn acceptor(&self, alpn_protocols: &[Vec<u8>]) -> Result<ssl::SslAcceptor> {
+        // mozilla_intermediate_v5 is the only variant that enables TLSv1.3, so we use that.
+        let mut conn = ssl::SslAcceptor::mozilla_intermediate_v5(ssl::SslMethod::tls_server())?;
+
+        // Force use of TLSv1.3.
+        conn.set_options(ssl::SslOptions::NO_TLSV1_2);
+        conn.clear_options(ssl::SslOptions::NO_TLSV1_3);
+
+        let roots = self.root_store()?;
+        tracing::debug!(
+            roots = ?self
+                .base
+                .roots
+                .iter()
+                .filter_map(|c| super::fingerprint(&*c))
+                .collect::<Vec<_>>(),
+            "Configuring acceptor roots",
+        );
+        conn.set_cert_store(roots);
+
+        // Ensure that client certificates are validated when present.
+        conn.set_verify(ssl::SslVerifyMode::PEER);
+
+        if let Some(certs) = &self.certs {
+            tracing::debug!(
+                cert = ?super::fingerprint(&*certs.leaf),
+                "Configuring acceptor certificate",
+            );
+            conn.set_private_key(&self.base.key)?;
+            conn.set_certificate(&certs.leaf)?;
+            conn.check_private_key()?;
+            for c in &certs.intermediates {
+                conn.add_extra_chain_cert(c.to_owned())?;
+            }
+        }
+
+        if !alpn_protocols.is_empty() {
+            let p = serialize_alpn(alpn_protocols)?;
+            conn.set_alpn_protos(&*p)?;
+        }
+
+        Ok(conn.build())
+    }
+
+    // TODO(ver) Specify certificate types, signing algorithms, cipher suites..
+    pub(crate) fn connector(&self, alpn_protocols: &[Vec<u8>]) -> Result<ssl::SslConnector> {
+        // XXX(ver) This function reads from the environment and/or the filesystem. This likely is
+        // at best wasteful and at worst unsafe (if another thread were to mutate these environment
+        // variables simultaneously, for instance). Unfortunately, the boring APIs don't really give
+        // us an alternative AFAICT.
+        let mut conn = ssl::SslConnector::builder(ssl::SslMethod::tls_client())?;
+
+        // Explicitly enable use of TLSv1.3
+        conn.set_options(ssl::SslOptions::NO_TLSV1 | ssl::SslOptions::NO_TLSV1_1);
+        // XXX(ver) if we disable use of TLSv1.2, connections just hang.
+        //conn.set_options(ssl::SslOptions::NO_TLSV1_2);
```

Comment:
should we report a bug for this upstream?

Reply:
maybe? this is probably a bug in boringssl and not the rust bindings, if it's a bug.